spec/helpers/samples_helper_spec.rb

```diff
@@ -0,0 +1,71 @@
+require "rails_helper"
+require "webmock/rspec"
+
+RSpec.describe SamplesHelper, type: :helper do
+  describe "#upload_samples_with_metadata" do
+    context "with basespace samples" do
+      before do
+        @project = create(:public_project)
+        @joe = create(:joe)
+        @host_genome = create(:host_genome)
+      end
+
+      let(:basespace_sample_attributes) do
+        [
+          {
+            basespace_access_token: "mock_access_token",
+            basespace_dataset_id: "mock_dataset_id",
+            host_genome_id: @host_genome.id,
+            name: "mock_sample",
+            project_id: @project.id
+          }
+        ]
+      end
+
+      let(:metadata_attributes) do
+        {
+          "mock_sample" => {
+            "Sample Type" => "CSF",
+            "Nucleotide Type" => "DNA",
+            "Collection Location" => "CA, USA",
+            "Collection Date" => "2010-01"
+          }
+        }
+      end
+
+      it "saved successfully" do
+        response = helper.upload_samples_with_metadata(
+          basespace_sample_attributes,
+          metadata_attributes,
+          @joe
+        )
+
+        expect(response["samples"].length).to be 1
+        expect(response["errors"].length).to be 0
+
+        created_sample = response["samples"][0]
+
+        expect(created_sample.name).to eq("mock_sample")
+        expect(created_sample.host_genome.id).to be @host_genome.id
+        expect(created_sample.project.id).to be @project.id
+        expect(created_sample.basespace_dataset_id).to eq("mock_dataset_id")
+        expect(created_sample.basespace_access_token).to eq("mock_access_token")
+
+        expect(Metadatum.where(sample_id: created_sample.id).length).to be 4
+      end
+
+      it "fails if neither basespace_access_token nor input_files_attributes is provided" do
```

Comment:
`input_file_attributes` has a very specific meaning in Rails. It's a [nested attribute](https://api.rubyonrails.org/classes/ActiveRecord/NestedAttributes/ClassMethods.html).

Since I don't want `basespace_access_token` to be inserted into a new InputFile model, I have to use a separate variable.